docs/proposals/ingesters-migration.md

```diff
@@ -0,0 +1,69 @@
+---
+title: "Migrating ingesters from chunks to blocks and back."
+linkTitle: "Migrating ingesters from chunks to blocks and back."
+weight: 1
+slug: ingesters-migration
+---
+
+- Author: @pstibrany
+- Reviewers:
+- Date: June 2020
+- Status: Proposed
+
+# Migrating ingesters from chunks to blocks
+
+This short document describes the first step in full migration of the Cortex cluster from using chunks storage to using blocks storage, specifically switching ingesters to using blocks, and modification of queriers to query both chunks and blocks storage.
+
+## Ingesters
+
+When switching ingesters from chunks to blocks, we need to consider the following:
+
+- Ingesting of new data, and querying should work during the switch.
+- Ingesters are rolled out with new configuration over time. There is overlap: ingesters of both kinds (chunks, blocks) are running at the same time.
+- Ingesters using WAL don’t flush in-memory chunks to storage on shutdown.
+- Rollout should be as automated as possible.
+
+How do we handle ingesters with WAL? There are several possibilities, but the simplest option seems to be adding a new flag to ingesters to flush chunks on shutdown. This is trivial change to ingester, and allows us to do automated migration by:
```

Comment:
Or something simpler, just don't send any data for 30mins (idle-timeout) before shutting down :)

Reply:
Oh wait, both the ingester deployments will share the same ring? That makes sense. Can we call out that we will be using the same statefulset and not going to create a new one for blocks? 

We also have a shutdown endpoint on the chunks ingester which just flushes and shuts down.

Reply:
> We also have a shutdown endpoint on the chunks ingester which just flushes and shuts down.

Yes we do, but I don't quite see how to use it for automation. On calling /shutdown, ingester flushes everything and stops. Kubernetes will just restart it. But we need restart with new configuration.

Reply:
I think it idles out, and doesn't stop iirc. @codesome ?

Reply:
@gouthamve you are right. Kubernetes won't restart it. /shutdown endpoint will flush all the chunks and remove itself from the ring and stay idle.

Reply:
> 4. Mention how the potential schema config will look like, specially in the case of a rollback. Currently we support splitting it by the UTC day boundary, but how will that work because the rollout won't align to that. I see that for a particular time-range we need to query both.

Schemas are low-level thing for configuring chunks store. We will not use schema for anything in this proposal.

Querier needs single timestamp to decide whether to query chunks store or not. Blocks store can always be queried, because querier knows whether to hit any block based on block metadata.

See my PR #2747 which implements this already.

I've updated "Querying" section of the document to better explain this.